faker.go

```diff
@@ -396,6 +397,57 @@ func FakeDataSkipFields(a interface{}, fieldsToSkip []string) error {
 	return nil
 }
 
+type generator struct {
+	index int
+}
+
+func (g *generator) Next(a interface{}) (done bool, err error) {
+	reflectType := reflect.TypeOf(a)
+
+	if reflectType.Kind() != reflect.Ptr {
+		return false, errors.New(ErrValueNotPtr)
+	}
+
+	if reflect.ValueOf(a).IsNil() {
+		return false, fmt.Errorf(ErrNotSupportedPointer, reflectType.Elem().String())
+	}
+
+	rval := reflect.ValueOf(a)
+
+	// if a is a pointer, set seenPointers to -1 so that the root itself is not set to nil
+	t := reflect.TypeOf(a)
+	seenPointers := 0
+	if t != nil {
+		k := t.Kind()
+		switch k {
+		case reflect.Ptr:
+			seenPointers = -1
+		}
+	}
+
+	finalValue, err := getValue(a, &seenPointers, g.index)
+	if err != nil {
+		return false, err
+	}
+	g.index++
+
+	// we are done when we have exhausted all pointers
+	done = g.index >= seenPointers
+	if !finalValue.IsNil() {
+		rval.Elem().Set(finalValue.Elem().Convert(reflectType.Elem()))
+	}
+	return done, nil
+}
+
+// FakeDataWithNilPointerGenerator returns a generator that can be used to create successive fakes with
+// a single pointer set to nil. This can be used to iterate through fakes with all possible pointer fields
+// set to nil, to catch nil pointer-dereference errors.
+func FakeDataWithNilPointerGenerator() *generator {
```

Comment:
Exported method returning an unexported struct looks weird

Reply:
Good point. It's technically valid but can prevent manipulation by packages using the function. We actually only need the `Next()` method, so I've changed it to return a public interface that exposes only that method instead.